Add support for STM32G474 family

hartmannathan opened the pull request:
## Summary

Add support for the STM32G474 family of microcontrollers and the B-G474E-DPOW1 Discovery Board, which features a STM32G474RET6.

## Impact

This is a major pull request as it adds support for an entirely new family of STM32. This support is implemented in arch/arm/src/stm32 and shares implementation with other STM32 families supported by that code, such as the 'L15xx, 'F10xx, 'F20xx, 'F3xxx, and 'F4xxx.

## Testing

The B-G474E-DPOW1 Discovery Board boots successfully through to the NSH prompt. NSH runs and is responsive.

hartmannathan commented:
### Note about nxstyle:

The pre-checks are going to fail on nxstyle errors. I am aware of this and am working to resolve the nxstyle issues. In particular, I am going to put the nxstyle fixes into a separate commit, because they are going to change a lot of unrelated lines that will make reviewing more difficult.

davids5 commented:
@hartmannathan - I will need the weekend to give this mind share.

hartmannathan commented:
> @hartmannathan - I will need the weekend to give this mind share.

Thanks! Take your time...

raiden00pl commented on arch/arm/include/stm32/stm32g47xxx_irq.h:
```diff
@@ -0,0 +1,204 @@
+#define STM32_IRQ_DMA1CH5     (STM32_IRQ_FIRST+15)  /* 15: DMA1 channel 5 global interrupt */
+#define STM32_IRQ_DMA1CH6     (STM32_IRQ_FIRST+16)  /* 16: DMA1 channel 6 global interrupt */
+#define STM32_IRQ_DMA1CH7     (STM32_IRQ_FIRST+17)  /* 17: DMA1 channel 7 global interrupt */
+#define STM32_IRQ_ADC1        (STM32_IRQ_FIRST+18)  /* 18: ADC1 and ADC2 global interrupt (see also STM32_IRQ_ADC2) */
```
This should be called STM32_IRQ_ADC12. Otherwise, it will not work with the current ADC implementation.

hartmannathan replied:
Good catch! Fixed in db33458048443413ea95313d3959ace7fb6e5183. Thank you.

raiden00pl commented on arch/arm/include/stm32/stm32g47xxx_irq.h:
```diff
@@ -0,0 +1,204 @@
+#define STM32_IRQ_NEXTINT     (102)
+#define NR_IRQS               (STM32_IRQ_FIRST+102)
+
+/* Interrupt vectors that are shared with numbers defined above: */
```
I think we should organize the file like other STM32 chips. So if we have a IRQ number shared between peripherals, we put the definitions one by one.

hartmannathan replied:
Done in fade67cca2d31a39ad8e746da83b5c65e57e1add.

raiden00pl commented on arch/arm/src/stm32/Kconfig:
```diff
@@ -1129,15 +1129,80 @@ config ARCH_CHIP_STM32F469N
 	select STM32_STM32F469
 	select STM32_HAVE_ETHMAC
 
+config ARCH_CHIP_STM32G474C
```
It seems to me that this is not correct. You should use chip names from [here](https://www.st.com/content/ccc/fragment/product_related/line_information/line_level_diagram/group0/cc/a2/82/c1/1b/2a/45/35/stm32g4x4_line_ln2130/files/stm32g4x4_line_ln2130.jpg/_jcr_content/translations/en.stm32g4x4_line_ln2130.jpg) (eg. STM32G474**CE**, STM32G474**CC** ... not  STM32G474**C**)
Each chip version should be a single entry in the "STM32 Chip Selection" choice menu.
Also I don't think that we need a distinction between LQFP/UFQFPN and LQFP/BGA packages. 
This does'nt affect OS in any way

hartmannathan replied:
I'm working on this change...

hartmannathan replied:
If I do what you suggest, the second letter determines the flash size:

STM32G474x**E** = 512Kbyte
STM32G474x**C** = 256Kbyte
STM32G474x**B** = 128Kbyte

How do I handle that? Like this:

```suggestion
config ARCH_CHIP_STM32G474CB
	bool "STM32G474CB"
	select STM32_STM32G47XX
	select STM32_STM32G474C
	select STM32_FLASH_CONFIG_B
```

Note that STM32_FLASH_CONFIG_B is part of a 'choice' and I don't see any other chip doing it that way.

hartmannathan replied:
> Also I don't think that we need a distinction between LQFP/UFQFPN and LQFP/BGA packages.
> This does'nt affect OS in any way

I removed distinction between LQFP/UFQFPN and LQFP/BGA packages in cc206e81b3b3d920c8031c025cbf6a368325febc.

raiden00pl replied:
Looks good. 
But I'm not sure if we need additional options for STM32_STM32G474M, STM32_STM32G474R, etc.
Shouldn't we keep chip-specific options just under ARCH_CHIP_STM32G474M, ARCH_CHIP_STM32G474R, etc ?
Then leave only STM32_STM32G47XX as the common configuration for the G47XX familly.
There is currently no consistency regarding this in Kconfig

hartmannathan replied:
> Looks good.
> But I'm not sure if we need additional options for STM32_STM32G474M, STM32_STM32G474R, etc.
> Shouldn't we keep chip-specific options just under ARCH_CHIP_STM32G474M, ARCH_CHIP_STM32G474R, etc ?
> Then leave only STM32_STM32G47XX as the common configuration for the G47XX familly.
> There is currently no consistency regarding this in Kconfig

Yes, Kconfig is inconsistent. I tried to follow what most other chips were doing.

What would your preference be?

raiden00pl replied:
I don't have any strong opinion about it, so let's leave it as it is.

hartmannathan replied:
OK

raiden00pl commented on arch/arm/src/stm32/hardware/stm32_dma_v1.h:
```diff
@@ -64,6 +64,9 @@
 #define DMA_CHAN5                  (4)
 #define DMA_CHAN6                  (5)
 #define DMA_CHAN7                  (6)
+#if defined(CONFIG_STM32_STM32G47XX)
```
It's better to introduce a new definition in this file like this 

#if defined(CONFIG_STM32_STM32G47XX)
#define HAVE_DMA_CHAN8 
#endif

We should avoid chip-specific #ifdef's if possible.

hartmannathan replied:
This is done in commit 710da07dea7a8ac40c00e1093cb697e63d05fb2a. Please see if that is satisfactory.

raiden00pl commented on arch/arm/src/stm32/stm32_dma_v1.c:
```diff
@@ -106,80 +114,132 @@ struct stm32_dma_s
+    .irq      = STM32_IRQ_DMA1CH8,
+    .base     = STM32_DMA1_BASE + STM32_DMACHAN_OFFSET(7),
+  },
+#endif /* DMA1_NCHANNELS > 7 */
```
Is there a reason why these conditions are nested? It looks strange :)

hartmannathan replied:
Fixed in 71070d32dacc8c39f104f3f9faa71280704c0d51.

raiden00pl commented on arch/arm/src/stm32/hardware/stm32g47xxx_memorymap.h:
```diff
@@ -0,0 +1,194 @@
+                                            /* 0xc0000000-0xdfffffff: Reserved */
+#define STM32_CORTEX_BASE    0xe0000000     /* 0xe0000000-0xffffffff: 512Mb Cortex-M4 block */
+
+#define STM32_REGION_MASK    Warning Not Cleared!  0xf0000000
```
#define STM32_REGION_MASK    0xf0000000
is OK.

hartmannathan replied:
Thank you!

Fixed in commit 8266c945d93cd0282dc7b8acbb0305f45933521e.

raiden00pl commented on arch/arm/src/stm32/hardware/stm32g47xxx_memorymap.h:
```diff
@@ -0,0 +1,194 @@
+#define STM32_CORTEX_BASE    0xe0000000     /* 0xe0000000-0xffffffff: 512Mb Cortex-M4 block */
+
+#define STM32_REGION_MASK    Warning Not Cleared!  0xf0000000
+#define STM32_IS_SRAM(a)     Warning Not Cleared!  ((((uint32_t)(a)) & STM32_REGION_MASK) == STM32_SRAM_BASE)
```
#define STM32_IS_SRAM(a)    ((((uint32_t)(a)) & STM32_REGION_MASK) == STM32_SRAM_BASE)

hartmannathan replied:
Thank you!

Fixed in commit 8266c945d93cd0282dc7b8acbb0305f45933521e.

davids5 commented on arch/arm/src/stm32/hardware/stm32g47xxx_pinmap.h:
```diff
@@ -0,0 +1,2627 @@
+
+/* USB Device Full Speed ****************************************************************************/
+
+#endif /* __ARCH_ARM_SRC_STM32_HARDWARE_STM32G47XXX_PINMAP_H */
```
You can uses something like `sunifdef` to fix this,

davids5 commented on arch/arm/src/stm32/hardware/stm32g47xxx_pwr.h:
```diff
@@ -0,0 +1,485 @@
+
+/* PWR Power Status Reset Register (SCR) */
+
+#define PWR_SCR_CSBF_SHIFT      (8)                                       /* Clear Standby Flag */
```
line then up please

hartmannathan replied:
Fixed in 076cceb7b7da7510a6c7aed40cae7789343956c0.

davids5 commented on arch/arm/src/stm32/hardware/stm32g47xxx_rcc.h:
```diff
@@ -0,0 +1,677 @@
+#define RCC_PLLCFGR_RESET              0x00001000
+
+/* CIER - Clock Interrupt Enable Register */
+
```
I try to do this files as a representation of the reference manual at the version I am working off of. It is a PITA but keeping the names and the comments consistent with the manual help to speed development and catch the bugs as the docs evolve. "Oh this is what we thought it does, but now the new rev of the manual says it does this"

hartmannathan replied:
Agreed, and I tried to do the same wherever possible. What isn't consistent with the manual here?

davids5 commented on arch/arm/src/stm32/hardware/stm32g47xxx_uart.h:
```diff
@@ -0,0 +1,420 @@
+ * Public Function Prototypes
+ ****************************************************************************/
+
+#endif /* __ARCH_ARM_SRC_STM32_HARDWARE_STM32G47XXX_UART_H */
```
Wow Nathan! - you did a  a lot of work. Stopping here for now, will do more later.

hartmannathan commented on arch/arm/src/stm32/hardware/stm32_adc_v2.h:
```diff
@@ -86,18 +86,24 @@
-#define STM32_ADC4_BASE              (STM32_ADC4_OFFSET + STM32_ADC34_BASE) /* ADC4 Slave ADC */
-#define STM32_ADC12CMN_BASE          (STM32_ADCCMN_OFFSET + STM32_ADC12_BASE) /* ADC1, ADC2 common */
-#define STM32_ADC34CMN_BASE          (STM32_ADCCMN_OFFSET + STM32_ADC34_BASE) /* ADC3, ADC4 common */
+#if !defined(CONFIG_STM32_STM32G47XX)
```
Can you tell me why the ADC driver is different from all other drivers? For all other drivers, peripheral base addresses are defined in the memorymap header.

It is less work to define ADC base addresses for G4 in stm32_adc_v2.h like you suggest. But I think it is more consistent with other drivers to move them all to the memorymap headers, unless there is a good reason not to.

I'll take your suggestion and move the defines to stm32_adc_v2.h for now, but we should consider moving all of them to the memorymap header in a later PR.

raiden00pl replied:
If you look at the memory map in reference manual, you'll see that the ADCs are grouped into blocks. We have only two independent ADC blocks: ADC12 and ADC345. This means that ADC1 and ADC2 (or ADC3, ADC4 and ADC5) are not separate units that can be controlled independently, but they are closely related. So basically there is no peripherals such as ADC1 or ADC2, but it is ADC12 and this should be reflected in the memory map.

hartmannathan replied:
That is OK, but why are the base address defines supposed to be in stm32_adc_v2.h and not in the memory map headers?

raiden00pl replied:
ADCx and ADCy offsets in the ADCxy memory region is an ADC IP block property, not a memory map property (and this approach matches the memory map table from the ST reference manual ;) )

hartmannathan replied:
OK this makes sense now. I will move these defines.

hartmannathan replied:
Done in 8c037c426a1d4c1e9130d710ec38ed69234c336c.

hartmannathan commented on arch/arm/src/stm32/hardware/stm32g47xxx_pinmap.h:
```diff
@@ -0,0 +1,2627 @@
+#define GPIO_ADC1_IN14                 (GPIO_ANALOG|GPIO_PORTB|GPIO_PIN11)
+#define GPIO_ADC1_IN15                 (GPIO_ANALOG|GPIO_PORTB|GPIO_PIN0)
+
+#if defined(CONFIG_ARCH_CHIP_STM32G474M) || \
```
@davids5, let's make sure I understand your suggestion:

We would have several files:

- stm32g47mxx_pinmap.h
- stm32g47cxx_pinmap.h
- etc.

Each of those files is a complete pinmap for that variant.

Then, stm32g47xxx_pinmap.h would choose which one to include.

hartmannathan replied:
I went ahead and made this change in cc206e81b3b3d920c8031c025cbf6a368325febc. The pinmap is much simplified.

hartmannathan commented on arch/arm/src/stm32/stm32_lowputc.c:
```diff
@@ -235,12 +235,18 @@
        USART_CR1_PCE | USART_CR1_WAKE | USART_CR1_M | USART_CR1_MME | \
        USART_CR1_OVER8 | USART_CR1_DEDT_MASK | USART_CR1_DEAT_MASK | \
        USART_CR1_ALLINTS)
+#  elif defined(CONFIG_STM32_STM32G47XX)
```
Fixed in 0600f8a4ab7d04b75e3a40d00a0766d087d77dae.

hartmannathan commented on arch/arm/include/stm32/stm32g47xxx_irq.h:
```diff
@@ -0,0 +1,204 @@
+ * Other interrupts (vectors >= 16) are defined below.
+ */
+
+#define STM32_IRQ_WWDG        (STM32_IRQ_FIRST+0)   /* 0:  Window Watchdog interrupt */
```
Yes, it is. I think it is not checking for operator spacing in defines.

I did it this way because looking at other code, it seems that's how it is being done elsewhere.

WDYT?

hartmannathan replied:
I went ahead and fixed operator spacing. All binary operators are surrounded by a single space per the coding standard.

hartmannathan commented on arch/arm/include/stm32/stm32g47xxx_irq.h:
```diff
@@ -0,0 +1,204 @@
+#define STM32_IRQ_FDCAN3_1    (STM32_IRQ_FIRST+89)  /* 89: FDCAN3 interrupt 1 */
+
+#define STM32_IRQ_RNG         (STM32_IRQ_FIRST+90)  /* 90: RNG global interrupt */
+#define STM32_IRQ_LPUART1     (STM32_IRQ_FIRST+91)  /* 91: LPUART1 */
```
Good catch! Fixed in d8f29f6263e4d22f0c3eb15547f14bae6f890eb2

hartmannathan replied:
> @hartmannathan - I can look at it again this weekend.

@davids5 - Thanks! Once you finish reviewing, assuming it looks good, feel free to remove the [DO NOT MERGE] from the title and do the honors!

I've confirmed that it still builds, flash-programs, and runs on the board, following all the cumulative changes made since I opened the PR. It's already rebased on latest master (as of this writing)...

Thanks for all your work so far!

hartmannathan commented on arch/arm/src/stm32/Kconfig:
```diff
@@ -1810,6 +1876,100 @@ config STM32_STM32F469
+	select STM32_HAVE_DAC4
+	select STM32_HAVE_FSMC
+	select STM32_HAVE_FMAC
+	select STM32_HAVE_FDCAN
```
Good catch! Fixed in 6cae597457f5d0ceca7fce23d43cfb1805fa99be.

hartmannathan commented on arch/arm/src/stm32/hardware/stm32g47xxx_memorymap.h:
```diff
@@ -0,0 +1,194 @@
+                                            /* 0x50060c00-0x5fffffff: Reserved */
+
+/* APB1 Base Addresses ******************************************************************************/
+
```
Yup, I was planning to, but forgot...

hartmannathan replied:
Fixed in c293babe6827d9f9c45dd2781fdeb1d167d1a814.

hartmannathan commented on arch/arm/src/stm32/Kconfig:
```diff
@@ -2577,7 +2863,7 @@ config STM32_SPI6
 	bool "SYSCFG"
 	default y
-	depends on STM32_STM32L15XX || STM32_STM32F30XX || STM32_STM32F33XX || STM32_STM32F37XX || STM32_STM32F20XX || STM32_STM32F4XXX || STM32_CONNECTIVITYLINE
+	depends on STM32_STM32L15XX || STM32_STM32F30XX || STM32_STM32F33XX || STM32_STM32F37XX || STM32_STM32F20XX || STM32_STM32F4XXX || STM32_STM32G47XX || STM32_CONNECTIVITYLINE
```
Should this be a separate PR after we complete this one? I think we should not pull in too many unrelated changes in one huge commit.

hartmannathan commented:
When I finish making all the changes you requested, will it be easier for you if I squash the commits? Or leave them separate?

davids5 commented:
> When I finish making all the changes you requested, will it be easier for you if I squash the commits? Or leave them separate?


@hartmannathan - you can use [mv-changes](https://github.com/MikhailGolubtsov/git-mv-changes)  ti regroup by subject (i.e. pinmap, adc) or squash by file.

hartmannathan commented:
> > When I finish making all the changes you requested, will it be easier for you if I squash the commits? Or leave them separate?
> 
> @hartmannathan - you can use [mv-changes](https://github.com/MikhailGolubtsov/git-mv-changes) to regroup by subject (i.e. pinmap, adc) or squash by file.

@davids5 -- ok... I'll wait until you and @raiden00pl tell me that you're ready, because rearranging commits will make it harder to see which change is related to which comment above...

Meanwhile I found and fixed some other issues, like filling in the missing ADC EXTSEL/JEXTSEL defines.

hartmannathan commented on arch/arm/src/stm32/hardware/stm32g47xxx_gpio.h:
```diff
@@ -0,0 +1,340 @@
+
+#define GPIO_MODER_SHIFT(n)            ((n) << 1)
+#define GPIO_MODER_MASK(n)             (0x3 << GPIO_MODER_SHIFT(n))
+
```
Done in 637afd3d4c5762f81db49cfce9f6e1b2527bf3a5.

davids5 commented on arch/arm/src/stm32/hardware/stm32_dma_v1.h:
```diff
@@ -54,6 +54,10 @@
 
+#if defined(CONFIG_STM32_STM32G47XX)
+#  define HAVE_DMA_CHAN8
+#endif
```
This should come in from chip.h or Kconfig.

hartmannathan replied:
> This should come in from chip.h or Kconfig.

I think there should there be **two** defines from Kconfig: HAVE_DMA1_CHAN8 and HAVE_DMA2_CHAN678.

hartmannathan replied:
Done in b5910332e199ae2d9aeabbe1e1341198f43f7b9c.

hartmannathan commented:
Woohoo! All checks have passed!

@davids5, @raiden00pl, please let me know if you're finished reviewing, or are continuing.

Meanwhile, I'm going to do one more build to verify it still works on hardware.

davids5 commented:
@hartmannathan - I can look at it again this weekend.

Ouss4 commented on boards/arm/stm32/b-g474e-dpow1/src/Make.defs:
```diff
@@ -0,0 +1,38 @@
+#
+############################################################################
+
+-include $(TOPDIR)/Make.defs
```
```suggestion
include $(TOPDIR)/Make.defs
```
To make it consistent with the recent changes to other makefiles. (#1088)

hartmannathan replied:
Good catch! Thank you reviewing and noticing that. I'll push a commit that fixes it soon...

hartmannathan commented:
> Looks good to me. Good job Nathan ;]

Thank you so much for taking the time to make your detailed review!!

hartmannathan commented:
FYI the above weirdness with changing base branch from 'master' to 'SocketCAN' and back to 'master' is because I force-pushed to the branch and GitHub did not update the PR. By changing the base branch, and changing it back, GitHub recognized the force push.

What changed in the force push:

- Rebased on latest master.
- Fixed some coding style issues found by davids5.
- Updated commit messages.
- No functional changes.

I confirmed that the board boots and runs with all the cumulative changes since I opened this PR.

davids5 commented:
> I confirmed that the board boots and runs with all the cumulative changes since I opened this PR.

Are there other boards this affects? Have you tested them as well?

hartmannathan commented:
> @hartmannathan - I can look at it again this weekend.

@davids5 - Thanks! Once you finish reviewing, assuming it looks good, feel free to remove the [DO NOT MERGE] from the title and do the honors!

I've confirmed that it still builds, flash-programs, and runs on the board, following all the cumulative changes made since I opened the PR. It's already rebased on latest master (as of this writing)...

Thanks for all your work so far!

hartmannathan commented:
> > I confirmed that the board boots and runs with all the cumulative changes since I opened this PR.
> 
> Are there other boards this affects? Have you tested them as well?

I don't have those boards. Do you mean build testing? I can build test some boards that use other processors in this directory.

hartmannathan commented:
> > > I confirmed that the board boots and runs with all the cumulative changes since I opened this PR.
> > 
> > 
> > Are there other boards this affects? Have you tested them as well?
> 
> I don't have those boards. Do you mean build testing? I can build test some boards that use other processors in this directory.

@davids5 , actually, I do have a nucleo-f302r8 board. The stm32 directory handles the f302r8 MCU. I'll go ahead and make sure that builds, boots, and runs...

hartmannathan commented:
> > > > I confirmed that the board boots and runs with all the cumulative changes since I opened this PR.
> > > 
> > > 
> > > Are there other boards this affects? Have you tested them as well?
> > 
> > 
> > I don't have those boards. Do you mean build testing? I can build test some boards that use other processors in this directory.
> 
> @davids5 , actually, I do have a nucleo-f302r8 board. The stm32 directory handles the f302r8 MCU. I'll go ahead and make sure that builds, boots, and runs...

I can confirm that the nucleo-f302r8:nsh configuration builds, flashes, and boots on the Nucleo F302R8 board.

Also, I have successfully built every "nsh" configuration under the boards/arm/stm32 directory. That is, I built these configurations and so I can confirm that my changes are not breaking these builds:

- axoloti:nsh
- b-g474e-dpow1:nsh
- clicker2-stm32:nsh
- cloudctrl:nsh
- fire-stm32v2:nsh
- hymini-stm32v:nsh
- maple:nsh
- mikroe-stm32f4:nsh
- nucleo-f103rb:nsh
- nucleo-f207zg:nsh
- nucleo-f302r8:nsh
- nucleo-f303re:nsh
- nucleo-f303ze:nsh
- nucleo-f334r8:nsh
- nucleo-f410rb:nsh
- nucleo-f429zi:nsh
- nucleo-f446re:nsh
- nucleo-l152re:nsh
- olimexino-stm32:nsh
- olimex-stm32-e407:nsh
- olimex-stm32-h407:nsh
- olimex-stm32-p107:nsh
- olimex-stm32-p207:nsh
- olimex-stm32-p407:nsh
- omnibusf4:nsh
- photon:nsh
- shenzhou:nsh
- stm3210e-eval:nsh
- stm3220g-eval:nsh
- stm3240g-eval:nsh
- stm32butterfly2:nsh
- stm32f103-minimum:nsh
- stm32f334-disco:nsh
- stm32f3discovery:nsh
- stm32f411e-disco:nsh
- stm32f429i-disco:nsh
- stm32f4discovery:nsh
- stm32ldiscovery:nsh
- stm32_tiny:nsh
- stm32vldiscovery:nsh
- viewtool-stm32f107:nsh

All of those configurations built successfully. There were no warnings pertaining to my changes.

This is the extent of testing that I can do on other boards/MCUs, because the only STM32 boards I have access to at this time are the B-G474E-DPOW1 Discovery and the Nucleo F302R8.

I was very careful when editing existing code and I'm confident that my changes are safe. Having said so, I encourage anyone and everyone who is interested in STM32 support to verify the code and to check that your boards and configurations continue to work with these changes applied. Our code quality is paramount.

@acassis you expressed interest in reviewing the code, do you still wish to look at it?

davids5 commented:
@acassis - Do you have time to take a pass? If not please remove yourself and I will merge it.

acassis commented:
Ok, go ahead!

hartmannathan commented:
Thanks to everyone for taking the time to review and for all your great finds and suggestions!
